accepted/2022/ReflectionInvoke.md

```diff
@@ -0,0 +1,474 @@
+
+Due to the performance and usability issues, workarounds and alternatives are used including:
+- [MethodInfo.CreateDelegate()](https://learn.microsoft.com/dotnet/api/system.reflection.methodinfo.createdelegate) or [`Delegate.CreateDelegate()`](https://learn.microsoft.com/dotnet/api/system.delegate.createdelegate) which supports a direct, fast method invocation. However, since delegates are strongly-typed, this approach does not work for loosely-typed invoke scenarios where the signature is not known at compile-time.
+- [`System.TypedReference`](https://learn.microsoft.com/dotnet/api/system.typedreference) along with `MakeTypedReference()` or `__makeref` can be used to modify a field or nested field directly. The `FieldInfo.SetValueDirect()` and `GetValueDirect()` can be used with `TypedReference` to get\set fields without boxing the value (but still boxes the target since that is still `object`).
+ - [Dynamic methods](https://learn.microsoft.com/dotnet/framework/reflection-and-codedom/how-to-define-and-execute-dynamic-methods) are used which are IL-emit based and require a non-trivial implementation for even simple things like setting property values. Those who use dynamic methods must have their own loosely-typed invoke APIs, which may go as far as using generics with `Type.MakeGenericType()` or `MethodInfo.MakeGenericMethod()` to avoid boxing. In addition, a fallback to standard reflection is required to support those platforms where IL Emit is not available.
+- Compiled [expression trees](https://learn.microsoft.com/dotnet/csharp/programming-guide/concepts/expression-trees/) which use dynamic methods if IL Emit is available but also conveniently falls back to standard reflection when not available. Using an expression to invoke a member isn't intuitive and brings along the large `System.Linq.Expressions.dll` assembly.
+
+# .NET 8 Goals
+The .NET 7 release had a 3-4x perf improvement for the existing `Invoke()` APIs by using IL Emit when available and falling back to standard reflection when IL Emit is not available. Although this improvement is significant, it still doesn't replace the need to use the IL-Emit based alternatives (dynamic methods and expression trees) for highly-performance-sensitive scenarios including the `System.Text.Json` serializer. New APIs are required that don't have the overhead of the existing `Invoke()` APIs.
+
+For .NET 8, there are two primary goals:
+1) Support byref-like types both for invoking and passing as arguments; this unblocks various scenarios. An unsafe approach may used for .NET 8 if support for `TypedReference` isn't addressed by Roslyn (covered later).
```

Comment:
I believe that we know by know that we are not going to get safe TypedReferences from Roslyn in .NET 8.

Reply:
Sure - I'll clarify that and move to the "future" section.

Reply:
Think there is a simple design where we have safe APIs for constructing `TypedReference` instances and getting the values back from them. The only type `unsafe` would come into the mix is when you attempted to extract the value of a `ref struct` by `ref`. That would put some burden on the implementation here to be careful with the values but consumers shouldn't need to. 

The conversation on whether or not this will work for the reflection APIs keeps lagging though so it's hard to tell if we should be investing in it.